### PR TITLE
Improve LSP1-UniversalReceiver + Delegate standard explanation

### PR DESCRIPTION
**Pull request by YamenMerhi**
## What does this PR introduce?
- Moving LSP1-UniversalReceiverDelegate to the generic standard section. Because LSP1-URD can be used in any SC not just UP.
### Before:
Screen Shot 2022-10-05 at 1 18 18 AM

### After:

Screen Shot 2022-10-05 at 1 18 43 AM

<br><hr>
- Add more content and figures to the introduction of the LSP1-UniversalReceiver Standard. + simplify the language

### Before:
Screen Shot 2022-10-05 at 1 19 42 AM

### After: 
Screen Shot 2022-10-05 at 1 20 01 AM
Screen Shot 2022-10-05 at 1 20 20 AM
Screen Shot 2022-10-05 at 1 20 36 AM

- Explain better + give examples in the specification + the extension (UniversalReceiverDelegate)
- Add the routing behaviour explanation in the LSP0-ERC725Account and LSP9-Vault

### CODE REVIEW

**Review by Hugoo — CHANGES_REQUESTED** [review]
Could you add a redirect in the Docusaurus config file? From old to new page? This is important as some websites may already link to this page:

**Comment by Hugoo** [comment]
https://docusaurus.io/docs/api/plugins/@docusaurus/plugin-client-redirects

**Review by b00ste — APPROVED** [review]

File: docs/standards/generic-standards/lsp1-universal-receiver-delegate.md [hunk 1]
```diff
@@ -27,23 +27,34 @@ Therefore, it is advised not to hardcode how the smart contract should handle an
...
 - address `caller`: the address that initially called the `universalReceiver(...)` function.
 
 - uint256 `value`: the amount of value sent to the `universalReceiver(...)` function.
 
 ### How Delegation works
 
-The address of the **[UniversalReceiverDelegate](../smart-contracts/lsp1-universal-receiver-delegate-up.md)** contract should be set as a value for the data key [`LSP1UniversalReceiverDelegate`](https://github.com/lukso-network/LIPs/blob/main/LSPs/LSP-0-ERC725Account.md#lsp1universalreceiverdelegate), inside the [ERC725Y Data key-value store](https://github.com/ERC725Alliance/erc725/blob/main/docs/ERC-725.md#erc725y) of the implementation contract (contract implementing the `universalReceiver(...)` function).
-
-With such an implementation, the `universalReceiver(...)` function can forward the call to its similar one called `universalReceiverDelegate(...)`.
+The address of the **[UniversalReceiverDelegate](../smart-contracts/lsp1-universal-receiver-delegate-up.md)** contract can be a variable inside the storage, to allow future upgrade of the **UniversalReceiverDelegate** simply by changing the address to another UniversalReceiverDelegate containing a new logic.
```
Comment (CJ42) [inline docs/standards/generic-standards/lsp1-universal-receiver-delegate.md:44]: _Suggested rewording:_ (not sure if too wordy, but my ideas was to keep it generic by saying _"inside the contract storage"_ instead of a _"variable"_).

```suggestion
The address of the **[UniversalReceiverDelegate](../smart-contracts/lsp1-universal-receiver-delegate-up.md)** contract can be stored inside the storage of the contract delegating its `universalReceiver(...)` functionality. This allows the upgrade of the **UniversalReceiverDelegate** simply by changing the address (in storage) to another UniversalReceiverDelegate containing a new logic.
```

File: docs/standards/generic-standards/lsp1-universal-receiver-delegate.md [hunk 2]
```diff
@@ -27,23 +27,34 @@ Therefore, it is advised not to hardcode how the smart contract should handle an
...
 
-This external contract will then implement the `universalReceiverDelegate(...)` function. It is also recommended that this contract registers the **[LSP1UniversalReceiverDelegate interfaceId](../smart-contracts/interface-ids.md)** using ERC165.
+If the contract implementing the `universalReceiver(..)` supports **[ERC725Y Data key-value store](https://github.com/ERC725Alliance/erc725/blob/main/docs/ERC-725.md#erc725y)**, the address of the **external contract** MUST be set as a value for the **LSP1UniversalReceiverDelegate data key** shown below to enable the optional extension. This key-value pair will act as a reference, making this external contract upgradeable if required.
```
Comment (CJ42) [inline docs/standards/generic-standards/lsp1-universal-receiver-delegate.md:46]: Suggesting to replace _"as shown below"_ by an anchor link (replace `#anchor-link` by the right anchor link in the page :)

```suggestion
If the contract implementing the `universalReceiver(..)` function supports **[ERC725Y Data key-value store](https://github.com/ERC725Alliance/erc725/blob/main/docs/ERC-725.md#erc725y)**, the address of the **external contract** MUST be set as a value under the **[`LSP1UniversalReceiverDelegate`](#anchor-link) data key**  to enable the optional extension. This key-value pair will act as a reference, making this external contract upgradeable if required.
```

File: docs/standards/generic-standards/lsp1-universal-receiver.md [hunk 3]
```diff
@@ -12,13 +12,23 @@ sidebar_position: 1
 :::
 
 ## Introduction
-There is often the need for smart contracts to **be aware of incoming transactions**, especially when it comes to value transfers.
+There is often the need for smart contracts to **be aware of incoming transactions**, especially when it comes to token transfers.
 
-A good example is ERC20 token transfers. When a smart contract receives a token, it has no generic way to be notified about it. One solution to this problem could be to monitor the receiving contract by listening for ERC20 token transfer events.
+A good example is **[ERC20](https://eips.ethereum.org/EIPS/eip-20)** token transfers. When a smart contract receives a token, it has no generic way **to be notified** about it. As the ERC20 token contract acts like a registry and during the transfer, the sender's balance is decreased, and the recipient's balance is increased **without further interaction**.
```
Comment (CJ42) [inline docs/standards/generic-standards/lsp1-universal-receiver.md:17]: _Suggesting splitting in multiple sentences_

```suggestion
A good example is **[ERC20](https://eips.ethereum.org/EIPS/eip-20)** token transfers. When a smart contract receives a token, it has no generic way **to be notified** about it. During the token transfer, the sender's balance decreases, and the recipient's balance increases. There are **no further interactions**, and the ERC20 token contract just acts as a registry.
```

File: docs/standards/generic-standards/lsp1-universal-receiver.md [hunk 4]
```diff
@@ -35,26 +45,50 @@ This standard defines a single function named `universalReceiver(...)` that coul
...
 - bytes `data`: Any arbitrary data.
 
-The `universalReceiver(...)` function emits an event with the data passed to it and some additional data. The function can then implement custom logic to make the contract behave differently based on the data received. For instance, the universalReceiver(...) function offers the following possibilities:
+> Receiving contracts should consider the `typeId` parameter to **decode the data correctly**.
+
+The `universalReceiver(...)` function **emits an event with the data passed to it and some additional data**. 
+
+![universalReceiver function emits an event](/img/standards/lsp1/universal-receiver-event.jpeg)
+
+For example, **to mark the receiving of a specific token**, during a token transfer, the token contract can call the `universalReceiver(..)` function of the token recipient with:
```
Comment (CJ42) [inline docs/standards/generic-standards/lsp1-universal-receiver.md:54]: _Suggested changes_

```suggestion
For example, **to notify the recipient that he is about to receive tokens**, during a token transfer, the token contract can call the `universalReceiver(..)` function of the recipient with:
```

File: docs/standards/generic-standards/lsp1-universal-receiver.md [hunk 5]
```diff
@@ -35,26 +45,50 @@ This standard defines a single function named `universalReceiver(...)` that coul
...
+
+- bytes32 `typeId`: **Hash**('ERCXXXXTokenReceived')
+- bytes `data`: **packedData**(amount of token sent, the sender address, the block timestamp)
+
+In this way, instead of **listening to all the events of the token contrats on the network**, and checking which one of these transfers is relative to the recipient, users can listen to the **[UniversalReceiver](../smart-contracts/lsp0-erc725-account.md#universalreceiver-1)** event on the contract implementing the `universalReceiver(..)` and know the token transfer details.  
+
+
+Also, the function can then implement **custom logic** to make the contract behave differently based on the data received. For instance, the `universalReceiver(...)` function, if customized, can do the following:
```
Comment (CJ42) [inline docs/standards/generic-standards/lsp1-universal-receiver.md:62]: _Suggesting shortening sentence

```suggestion
The `universalReceiver(...)` function implement **custom logic** to make the contract behave differently based on the data received. Some ideas include:
```

File: docs/standards/generic-standards/lsp1-universal-receiver.md [hunk 6]
```diff
@@ -35,26 +45,50 @@ This standard defines a single function named `universalReceiver(...)` that coul
...
 
 - Reverting on calls to completely disallow the smart contract from receiving assets, information, etc. :x:
 - Registering the received assets inside the contract storage (see [LSP5 - Received Assets](../universal-profile/lsp5-received-assets.md)). :heavy_plus_sign:
```
Comment (CJ42) [inline docs/standards/generic-standards/lsp1-universal-receiver.md:65]: I suggest being as creative as possible here 🎨 🙂 , with multiple ideas to really explain **how extensible LSP1 is:**

- Disallow receiving specific tokens from specific token contract addresses, for instance (_e.g: spam tokens)._
- Forward and store specific tokens in a [LSP9 Vault]("link to LSP9 Standard").
- Stake-specific tokens in a staking contract (e.g: governance tokens).
- Forward specific tokens in a contract behind a protocol or dApp (e.g: liquidity or lending pool to earn interest).
- Depending on the `typeId`, save a percentage % of tokens received (native tokens or not), by placing them in an LSP9 Vault for instance.

File: docs/standards/generic-standards/lsp1-universal-receiver.md [hunk 7]
```diff
@@ -35,26 +45,50 @@ This standard defines a single function named `universalReceiver(...)` that coul
...
 
 :::info
 
-See the **[LSP1-UniversalReceiverDelegate](../universal-profile/lsp1-universal-receiver-delegate.md)** standard for more details.
+See the **[LSP1-UniversalReceiverDelegate](../generic-standards/lsp1-universal-receiver-delegate.md)** standard for more details.
 
 :::
 
-LSP1-UniversalReceiverDelegate is an **optional extension** to the LSP1-UniversalReceiver Standard. As well as notifying a contract about the incoming and outgoing transactions via an event, it will delegate the `universalReceiver(...)` functionality to an external contract that can **handle and react to specific calls** with its custom logic.
+Customizing the `universalReceiver(..)` function is an option that users have to allow the function to **behave differently on different actions**. However, it's not advised to hardcode the logic of reacting on specific actions inside the function, as **this logic may need to change in the future** depending on several factors (eg. the vault where the tokens are forwarded get hacked, new staking contract is deployed, decided to revert on specific tokens later).
```
Comment (CJ42) [inline docs/standards/generic-standards/lsp1-universal-receiver.md:80]: _Suggested changes_

```suggestion
Overriding and customizing the `universalReceiver(..)` function is an option for users to allow **different behaviours depending on the data received**. However, it's not advised to hardcode the logic of reacting to specific actions inside the function because **this logic may need to change in the future** depending on several factors (eg. the vault where the tokens are forwarded gets compromised, a new staking contract is deployed, decided to revert on specific tokens later). 
```

File: docs/standards/generic-standards/lsp1-universal-receiver.md [hunk 8]
```diff
@@ -35,26 +45,50 @@ This standard defines a single function named `universalReceiver(...)` that coul
...
+
+**[LSP1-UniversalReceiverDelegate](../generic-standards/lsp1-universal-receiver-delegate.md)** is an **optional extension** to the **[LSP1-UniversalReceiver](#)** standard. As well as notifying a contract about the incoming and outgoing transactions by emitting an event, it can delegate the call to an external contract that can **handle and react to specific calls** with its custom logic.
+
+![Universal Receiver Delegate contract](/img/standards/lsp1/universal-receiver-delegate.jpeg)
+
+
+The address of the **external contract** can be a variable that changes in the contract storage by a setter function. In this way, users can customize such contracts to implement a specific logic that is changeable anytime in the future.
```
Comment (CJ42) [inline docs/standards/generic-standards/lsp1-universal-receiver.md:87]: _Suggested changes_

```suggestion
The address of the **external contract** can be stored and changed inside the contract storage. This way, users can customize such contracts to implement a specific logic that is changeable at any time.
```

File: docs/standards/generic-standards/lsp1-universal-receiver.md [hunk 9]
```diff
@@ -35,26 +45,50 @@ This standard defines a single function named `universalReceiver(...)` that coul
...
+
+![Multiple Universal Receiver Delegate](/img/standards/lsp1/multiple-urd.jpeg)
 
-The address of the **external contract** MUST be set as a value for the **LSP1UniversalReceiverDelegate data key** shown below to enable the optional extension. This key-value pair inside the **[ERC725Y Data key-value store](https://github.com/ERC725Alliance/erc725/blob/main/docs/ERC-725.md#erc725y)** of the contract implementing the `universalReceiver(...)` function will act as a reference, making this external contract upgradeable if required.
+If the contract implementing the `universalReceiver(..)` supports **[ERC725Y Data key-value store](https://github.com/ERC725Alliance/erc725/blob/main/docs/ERC-725.md#erc725y)**, the address of the **external contract** MUST be set as a value for the **LSP1UniversalReceiverDelegate data key** shown below to enable the optional extension. This key-value pair will act as a reference, making this external contract upgradeable if required.
```
Comment (CJ42) [inline docs/standards/generic-standards/lsp1-universal-receiver.md:91]: Same changes suggested as above.

BUt if its's in two places and duplicated, maybe we can remove it in one of them?

File: docs/standards/universal-profile/lsp9-vault.md [hunk 10]
```diff
@@ -48,25 +48,43 @@ The data keys and values are constructed according to the **[LSP2-ERC725YJSONSch
...
+:::info
+
+See the **[LSP1-UniversalReceiver](../generic-standards/lsp1-universal-receiver.md)** standard for more information.
 
-:::note
-Check **[LSP1-UniversalReceiver](../generic-standards/lsp1-universal-receiver.md)** standard for more information.
 :::
 
-## Extension
+This standard enables the vault to be notified of incoming transactions such as token transfer, followers, information transfer, etc. Notifications are handy for situations where users want to customize how their account contract reacts to certain tokens by either rejecting them or operating a specific call on each token received.
```
Comment (CJ42) [inline docs/standards/universal-profile/lsp9-vault.md:63]:
```suggestion
This standard enables the vault to be notified of incoming transactions such as token transfers, information transfers, etc. Notifications are handy when users want to customize how their account contract reacts to certain tokens by either rejecting them or operating a specific call on each received token.
```

File: docs/standards/universal-profile/lsp9-vault.md [hunk 11]
```diff
@@ -48,25 +48,43 @@ The data keys and values are constructed according to the **[LSP2-ERC725YJSONSch
...
+  "name": "LSP1UniversalReceiverDelegate:<bytes32>",
+  "key": "0x0cfc51aec37c55a4d0b10000<bytes32>",
+  "keyType": "Mapping",
+  "valueType": "address",
+  "valueContent": "Address"
+}
+```
 
-Such functionality can be attached by setting an **[LSP1-UniversalReceiverDelegate](./lsp1-universal-receiver-delegate.md)** to your account.
+> <bytes32\> is the `typeId` passed to the `universalReceiver(..)` function.
```
Comment (CJ42) [inline docs/standards/universal-profile/lsp9-vault.md:93]: why is it surrounded by `<` `>`?

Comment (YamenMerhi) [reply docs/standards/universal-profile/lsp9-vault.md:93]: To mark that the <bytes32> in the schema

File: docs/standards/generic-standards/lsp1-universal-receiver.md (hunk 6 above)
Comment (YamenMerhi) [reply docs/standards/generic-standards/lsp1-universal-receiver.md:65]: Good ideas, will not specify LSP9Vault as it could be overcomplicated for new users reading the standard, will keep it generic and write vaults

File: docs/standards/generic-standards/lsp1-universal-receiver-delegate.md (hunk 2 above)
Comment (YamenMerhi) [reply docs/standards/generic-standards/lsp1-universal-receiver-delegate.md:46]: Not sure if i can link it because it's not part of a specific block, just text under the paragraph

File: docs/standards/generic-standards/lsp1-universal-receiver.md (hunk 9 above)
Comment (YamenMerhi) [reply docs/standards/generic-standards/lsp1-universal-receiver.md:91]: I would keep it, as people could read the 2 standards separately, it's not that big deal imo

File: docs/standards/generic-standards/lsp1-universal-receiver-delegate.md [hunk 12]
```diff
@@ -27,23 +27,34 @@ Therefore, it is advised not to hardcode how the smart contract should handle an
 
 ### Specification
 
-This standard represents a smart contract delegated to the initial `universalReceiver(...)` function.
+:::success recommendation
 
-It contains a single function named `universalReceiverDelegate(...)` that takes the same parameters as the `universalReceiver(...)` function with two additional params:
+Smart contracts implementing the `universalReceiverDelegate(...)` function SHOULD **register** the **[LSP1UniversalReceiverDelegate InterfaceId](../smart-contracts/interface-ids.md) using ERC165**. This way, other contracts can be aware that the contract supports the LSP1 standard.
```
[the remark this answers is not in the document]
Comment (YamenMerhi) [reply docs/standards/generic-standards/lsp1-universal-receiver-delegate.md:32]: Will go with standard

File: docs/standards/generic-standards/lsp1-universal-receiver-delegate.md (hunk 2 above)
Comment (CJ42) [reply docs/standards/generic-standards/lsp1-universal-receiver-delegate.md:53]: ah ok. or maybe if the _optional extension_ is a subheading?

**Comment by CJ42** [comment]
@YamenMerhi once you made the changes, don't forget to make your branch up to date with the `main` branch.

**Review by Hugoo — APPROVED** [review]
